Add integration tests to GitHub Actions

jissereitsma opened the pull request:
Showcase https://github.com/yireo/Yireo_Webp2/actions is already setup and should be re-run after merging, to confirm that this is working.

fooman commented:
hey @jissereitsma thanks for pushing this forward. I did have some plans to carry over some of this to ExtDN https://github.com/fooman/magento2-starter-template/blob/master/.github/workflows/deploy.yml#L6 but I am keen to have something to start with and then iterate on it.

I would suggest we merge this (needed already just to get the commit history) and then we make a few changes. Happy to PR those. Things that I believe would be beneficial:

1. Move the extension install prior to the Magento setup command (this will catch missing sequence information when someone tries to alter DB info that hasn't yet been set up)
2. Use a composer require instead of copying to app/code as otherwise any extensions with dependencies miss out. The action input might then change to 
MODULE_COMPOSER_NAME (or read from composer.json)  
MODULES_TO_ENABLE (which would allow for packages that include multiple modules)  
3. If you are happy with it we can prebuilt and host the Docker image under the [ExtDN](https://hub.docker.com/u/extdn) account. It would just need to be added here https://github.com/extdn/github-actions-m2/blob/master/.github/workflows/docker.yml#L15 and every time a release is tagged it runs automated push to docker hub
4. What do you think of letting the repo provide the phpunit file? Ie have a path to phpunit as input (with a default location)
5. Out of curiosity what is the memory file patch needed for?
6. I think we can simplify the DB setup to just
https://github.com/fooman/magento2-starter-template/blob/master/.github/workflows/deploy.yml#L16
as the integration tests run their own setup:install

jissereitsma commented:
1. Done.
2. Done. I've renamed the variables to be `COMPOSER_NAME` and `MODULE_NAME`. It simplifies things.
3. Agreed. But that requires also access to the ExtDNs Docker Hub account, right? Or am I misunderstanding this point?
4. I've added a new variable `PHPUNIT_FILE` which can be filled in. It is empty by default, to fall back to the file provided in the Docker image.
5. The patch fixes an issue with a wrong memory check by the testing framework, after all tests have successfully run. The check runs a certain `ps` command that runs fine on my own Linux environment. But within Docker environments (at least my own), this fails, which makes the testing framework retry the heck as if the current environment was a Windows environment (huh?). And that Windows check fails again with a message that `tasklist.exe` could not be found. A very strange issue. That's why I patched the `Memory` class to simply return the exception message instead of retrying under Windows.
6. I'm not sure if this makes it simpler. The integration tests require their own database, but they also require a valid `app/etc/config.php`, which is generated by running a regular install first in a separate database. Because of this, I have opted for a PHP script to create the two databases, instead of just using the `MYSQL_DATABASE` variable. I'm not sure what the added value of the Docker image `quay.io/warden/mysql` is? The regular `mysql` image works just fine and it does not seem that much of a difference?

fooman commented:
Good stuff. Just on 3.) this repo's secrets are already set up so that it is automated.
Will take another look at 6. - it wasn't so much about the image, but rather only having 1 database in total.
